### PR TITLE
Store cc & identities forms

### PR DESCRIPTION
**Pull request by GioSensation**
**Reviewer:** @shakyShane 
**Asana:** https://app.asana.com/0/0/1201749050559820/f

## Description
Adds storage capabilities for credit cards and identities forms. We require a minimum set of values for each data type in order to create a usable entry. Namely:

- For credentials:
  - Only the password is required
- For identities:
  - Full name (either First + Last, or single field, in this second case we would store it as first name)
  - At least partial address:
    - street
    - city
    - ~~country~~ I decided against requiring the country because getting the country code is less reliable and we risk getting too many failures. We can easily add it later.
- For credit cards:
  - Card number 
  - Security code

More info [here](https://app.asana.com/0/0/1201530813714191/f).

I haven't changed how the form submission detection works at all. We can follow up on this separately.

## Steps to test
Added a bunch of tests for reading the form values. They're not exhaustive but a good start.

You can't check the whole on the app because it needs native work to accept the new data type and react to it.

### CODE REVIEW

File: src/DeviceInterface/AppleDeviceInterface.js
Comment (GioSensation) [inline src/DeviceInterface/AppleDeviceInterface.js:79]: New API method. Not yet implemented in the native layer.

File: src/Form/Form.js
Comment (GioSensation) [inline src/Form/Form.js:24]: Added plural for consistency with other sets and with the API.

File: src/Form/Form.js
Comment (GioSensation) [inline src/Form/Form.js:81]: These are set to `null` if they don't have enough data to form a usable data entry.

File: src/Form/Form.test.js
Comment (GioSensation) [inline src/Form/Form.test.js:42]: You've got the code to fix this in your Branch.

File: src/Form/countryNames.js
Comment (GioSensation) [inline src/Form/countryNames.js:4]: I've switched to using the `Intl.DisplayNames` mainly because we use it for filling the data. If we didn't use the exact same set of names we might have ended up autofilling a field and the same field wouldn't be available for storage. It's an edge case, but this solves it.

For context, we only use this list when `Intl.DisplayNames` itself is not available in the browser.

File: src/Form/countryNames.js
Comment (GioSensation) [inline src/Form/countryNames.js:291]: Initially I was generated it on the fly, but there's no harm in keeping this hardcoded. Saves computation on each page load.

File: src/Form/Form.js
Comment (GioSensation) [inline src/Form/Form.js:57]: This is now done in `prepareFormValuesForStorage`.

File: src/Form/formatters.js
Comment (GioSensation) [inline src/Form/formatters.js:201]: We could get smarter with this, but I think this is good enough for this iteration. Let me know if you disagree.

File: src/Form/inputTypeConfig.js
Comment (GioSensation) [inline src/Form/inputTypeConfig.js:29]: This is the change I mentioned yesterday with the `readonly` attribute. We want to select them, but we don't want to decorate them (decorate here means adding icons and listeners).

File: src/Form/inputTypeConfig.js
Comment (GioSensation) [inline src/Form/inputTypeConfig.js:42]: This is added to all `shouldDecorate` functions. If you can think of a more elegant way of doing this without all the repetition, I'd be happy to change it.

File: src/device-interface.d.ts
Comment (GioSensation) [inline src/device-interface.d.ts:34]: This is the right name. I'm not sure when we added the wrong one.

File: src/device-interface.d.ts
Comment (GioSensation) [inline src/device-interface.d.ts:29]: Internal versions of these objects include other properties that we use for convenience. I've created the distinction so we're sure we're sending the right stuff over the API.

File: src/device-interface.d.ts
Comment (GioSensation) [inline src/device-interface.d.ts:38]: Note that these are actually `number`s in the API. I've asked the native team to change the implementation to `string` and have a tiny patch locally to make sure the logic keeps working.

File: src/Form/formatters.js
Comment (shakyShane) [reply src/Form/formatters.js:173]: Note: It may be that instead of `Partial`, you might want to change the raw definition of the type in question, to have optional fields instead? I'll let you make the decision on which :)

File: src/Form/Form.js
Comment (shakyShane) [reply src/Form/Form.js:81]: yeah - there's a note about increasing type-safety on this stuff below :)

File: src/Form/formatters.js
Comment (shakyShane) [reply src/Form/formatters.js:201]: to comment on this, I'd need to have some examples of values that `creditCards.expiration` might have?

File: src/device-interface.d.ts
Comment (shakyShane) [reply src/device-interface.d.ts:38]: normally I would say let's update the type to `string | number` in that case, because then Typescript can check that we don't accidentally assume it's one or the other (and cause exceptions with method calls etc)

but if you have this handled elsewhere and the change is in flight, then no change needed

**Review by shakyShane — CHANGES_REQUESTED** [review]
@GioSensation 👋 Looking good!

just a few minor points raised, mostly around leveraging Typescript where we can :)

File: src/Form/formatters.js
Comment (GioSensation) [reply src/Form/formatters.js:136]: Good catch. Unlikely to happen, but I will add `el.selectedOptions[0]?.text`.

File: src/Form/formatters.js
Comment (GioSensation) [reply src/Form/formatters.js:146]: Yep, good point. I'll do that.

File: src/Form/Form.js
Comment (GioSensation) [inline src/Form/Form.js:208]: We are now checking for likely submit buttons first to minimise risk of false positive submissions.

File: src/Form/Form.js
Comment (GioSensation) [inline src/Form/Form.js:68]: We now use `hasFocus` to check that the form is focused when we detect an `Enter` keypress and fire the `submitHandler`.

File: src/Form/formatters.js
Comment (GioSensation) [inline src/Form/formatters.js:14]: As discussed previously, we changed the API to always return strings.

File: src/Form/formatters.js
Comment (GioSensation) [inline src/Form/formatters.js:172]: This is the main change in this file since your previous review. We've added support for storing cc forms on Amazon. They don't have a cvv field by design so we were failing to capture forms there. With this change, if cvv is missing but we have everything else we still save.

File: src/Form/formatters.js
Comment (GioSensation) [reply src/Form/formatters.js:225]: There is the possibility of `YYYY / MM`, but it's definitely rare. We can follow up or I can add more logic if I have some downtime today.

File: src/Form/matching-configuration.js
Comment (GioSensation) [inline src/Form/matching-configuration.js:257]: This is to enable matching of `username:` we were failing because of the `:`.

File: src/Form/selectors-css.js
Comment (GioSensation) [inline src/Form/selectors-css.js:18]: Again, removing `placeholder` and `subject` from the match. Ultimately we may want to move away from these complex selectors and have a more robust logic with regexes. Too risky now.

File: src/scanForInputs.js
Comment (GioSensation) [inline src/scanForInputs.js:14]: These are not necessarily `form` elements, because sometimes no formal `form` is defined and we try to infer a container element. This type is more accurate.

File: src/Form/matching-configuration.js
Comment (shakyShane) [reply src/Form/matching-configuration.js:255]: This is interesting, perhaps forceUnknown should not require an initial match then - basically making `match` completely optional

File: src/Form/Form.js
Comment (shakyShane) [reply src/Form/Form.js:217]: **nit:**

@GioSensation let's not block on it right now of course, but we might be able to avoid the 2 separate ways of checking here - not for perf or anything like that, but just to clarify intent.

File: src/Form/matching-configuration.js
Comment (GioSensation) [reply src/Form/matching-configuration.js:255]: I tried that, but it made a lot of test fail. I'm looking into it again, because I think I know why the tests were failing (some `forceUnknown` should actually have been `skip`). Looking…

File: src/Form/formatters.js
Comment (shakyShane) [inline src/Form/formatters.js:173]: I think these chained conditions are at the point where they could benefit for being expanded out, something like 

```ts
const shouldStoreCreditCards = ({creditCards}) => {
    if (!creditCards.cardNumber) return false;
    if (creditCards.cardSecurityCode) return true;
    if (creditCards.cardName) {
        if (creditCards.expiration) return true;
        return creditCards.expirationYear && creditCards.expirationMonth;
    }
    return false;
}
```

should make it easier for others to visually parse/understand the logic here.

I don't want to block the PR, but let's make a follow-up on this.

File: src/Form/formatters.js
Comment (shakyShane) [inline src/Form/formatters.js:187]: **nit**

(again, not blocking)

Since we introduced the `InputTypes` folder, I believe that going forward we could start to use those files as a catch-all place for related logic. Meaning, if I want to see code relating to how Credentials are prepared, I'd go the `InputTypes/Credentials.js` file and expect some interface to be adhered to in there.

**Review by shakyShane — APPROVED** [review]
@GioSensation nothing blocking from me, just a couple of comments to consider for once this has landed :)

File: src/Form/matching-configuration.js
Comment (GioSensation) [reply src/Form/matching-configuration.js:255]: Yep, it worked. I've just pushed an update to check `forceUnknown` before anything else and updated the configs to use `forceUnknown` more accurately (sometimes it should have been `skip`). 🎉

File: src/Form/test-cases/index.js
Comment (shakyShane) [inline src/Form/test-cases/index.js]: 🎉

File: src/Form/formatters.js
Comment (GioSensation) [reply src/Form/formatters.js]: I've added tests and implemented your suggestion. Looks much better now!

File: src/Form/formatters.js
Comment (GioSensation) [reply src/Form/formatters.js:190]: We'll follow up on this.

File: src/Form/formatters.js
Comment (shakyShane) [inline src/Form/formatters.js:173]: nice!

File: src/Form/selectors-css.js
Comment (shakyShane) [reply src/Form/selectors-css.js:18]: agreed - a lot of the complexity in `matchers.js` is around supporting the 3 layers. It's something we can work to reduce

**Pull request merged**
